Review notes:

* I tested if it worked using the following code:

```
Python 3.12.5 | packaged by conda-forge | (main, Aug  8 2024, 18:36:51) [GCC 12.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import scipy
>>> data_input = scipy.io.mmread('/file/does/not/exist')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/nodell/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/__init__.py", line 360, in mmread
    cursor, stream_to_close = _get_read_cursor(source)
                              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/__init__.py", line 198, in _get_read_cursor
    raise FileNotFoundError(f"The source file does not exist: {path}")
FileNotFoundError: The source file does not exist: /file/does/not/exist
```

   ✅ 

* Since this check only exists in the branch where the source file is not a gz or bz2 archive, I additionally checked those.

```
>>> data_input = scipy.io.mmread('/file/does/not/exist.gz')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/nodell/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/__init__.py", line 360, in mmread
    cursor, stream_to_close = _get_read_cursor(source)
                              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/__init__.py", line 190, in _get_read_cursor
    source = gzip.GzipFile(path, 'r')
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/.miniforge3/envs/scipy-dev/lib/python3.12/gzip.py", line 192, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/file/does/not/exist.gz'
>>> data_input = scipy.io.mmread('/file/does/not/exist.bz2')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/nodell/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/__init__.py", line 360, in mmread
    cursor, stream_to_close = _get_read_cursor(source)
                              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/__init__.py", line 194, in _get_read_cursor
    source = bz2.BZ2File(path, 'rb')
             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/.miniforge3/envs/scipy-dev/lib/python3.12/bz2.py", line 81, in __init__
    self._fp = _builtin_open(filename, mode)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/file/does/not/exist.bz2'
```

   These also work, for the reason that the BZ2 and GZIP libraries provide good error messages in the case where they are called on a file that does not exist. No additional check is needed here.

* `tmpdir = mkdtemp(suffix=str(threading.get_native_id()))` This part looks strange to me. Typically there would be no need to add the thread number to the temporary directory, because each call to mkdtemp will result in a new file. But, I see that other places in the same test file do this, and it was added in 5b36615052e7e871103fae9601ecef6c80b97033 to support the free-threaded tests. I don't understand why this was done, but doing it consistently with the rest of the file is a good idea.
* `FileNotFoundError: The source file does not exist: {path}` Is this the best error message? It's different from the normal error you would get from opening a file, `FileNotFoundError: No such file or directory: 'test'`. However, the error message is clear, and references the 'source' argument to the mmread() function. Good overall.
* On a related note, this will show the same error message if path is not a file, and is a directory. The error message is still reasonably clear here, so this is fine.
* If the input path is a FIFO, then it is not a file, but we could still reasonably read from it. Here's a concrete example. Suppose we have this Python code:

```
import scipy
import sys

data_input = scipy.io.mmread(sys.argv[1])
print(data_input)
```

Then we run this like so:

```
python script.py gr_900_900_crg.mm
```

This runs fine, and prints out the matrix file.

If we were using process substitution, like this

```
python script.py <(cat gr_900_900_crg.mm)
```

then Python will receive a FIFO rather than a file. Prior to this PR, this succeeds, but after, it fails, with this error:

```
Traceback (most recent call last):
  File "/home/nodell/scipy/dev.py", line 1740, in <module>
    cli()
  File "/home/nodell/.miniforge3/envs/scipy-dev/lib/python3.12/site-packages/rich_click/rich_command.py", line 367, in __call__
    return super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/.miniforge3/envs/scipy-dev/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/.miniforge3/envs/scipy-dev/lib/python3.12/site-packages/rich_click/rich_command.py", line 152, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/nodell/.miniforge3/envs/scipy-dev/lib/python3.12/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/.miniforge3/envs/scipy-dev/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/.miniforge3/envs/scipy-dev/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/.miniforge3/envs/scipy-dev/lib/python3.12/site-packages/pydevtool/cli.py", line 271, in callback
    sys.exit(cls.run(**kwargs))
             ^^^^^^^^^^^^^^^^^
  File "/home/nodell/scipy/dev.py", line 1395, in run
    exec(script, ns)
  File "<string>", line 4, in <module>
  File "/home/nodell/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/__init__.py", line 360, in mmread
    cursor, stream_to_close = _get_read_cursor(source)
                              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nodell/scipy/build-install/lib/python3.12/site-packages/scipy/io/_fast_matrix_market/__init__.py", line 198, in _get_read_cursor
    raise FileNotFoundError(f"The source file does not exist: {path}")
FileNotFoundError: The source file does not exist: /dev/fd/63
```

This is potentially a regression in functionality we used to support. I say 'potentially' because I am not sure anyone really loads matrices from FIFOs.

--- a/scipy/io/_fast_matrix_market/__init__.py
+++ b/scipy/io/_fast_matrix_market/__init__.py
@@ -194,7 +194,7 @@ def _get_read_cursor(source, parallelism=None):
             source = bz2.BZ2File(path, 'rb')
             ret_stream_to_close = source
         else:
-            if not os.path.isfile(path):
+            if not os.path.exists(path):
                 raise FileNotFoundError(f"The source file does not exist: {path}")
             return _fmm_core.open_read_file(path, parallelism), ret_stream_to_close
 

